### DISCUSSION TITLE
Finite-state machine

### DISCUSSION THREAD

**Discussion by giicoo** [mymmrac/telego#125] [Q&A] [answered]
Can you tell me how to implement a finite-state machine using handlers or some other built-in tool?
I am writing a bot and I don’t understand how to optimally remember the sequence of answers and what stage the user is currently at.

**Comment by mymmrac** [comment, marked as answer]
Hi, unfortunately Telego doesn't provide FSM as of now (it may later, who knows) if you want to have such functionality you can come up with your own system storing user state in memory or in some DB, also you can search for some libraries that implement FSM

Some libraries that you may look into:
- https://github.com/looplab/fsm
- https://github.com/qmuntal/stateless
- https://github.com/hishamk/statetrooper

**Reply by giicoo** [reply]
Thanks

**Reply by mnogokotin** [reply]
@mymmrac Hi. Can you add an example? It's very common task. :)

**Reply by mymmrac** [reply]
@mnogokotin personally for my bots I always did something different based on requirements each time, seems like FSM is very common thing, but actually when it goes to implementation there are many many ways to do that, so unfortunately I can't provide some generic example on how to do that
